File: hooks/internal/hooks/checkconsolelog.go
package hooks

import (
	"regexp"
	"strings"

	"github.com/oguzsh/everything-claude-code/internal/fileutil"
	"github.com/oguzsh/everything-claude-code/internal/gitutil"
	"github.com/oguzsh/everything-claude-code/internal/hookio"
)

var excludedPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\.test\.[jt]sx?$`),
	regexp.MustCompile(`\.spec\.[jt]sx?$`),
	regexp.MustCompile(`\.config\.[jt]s$`),
	regexp.MustCompile(`scripts/`),
	regexp.MustCompile(`__tests__/`),
	regexp.MustCompile(`__mocks__/`),
}

// CheckConsoleLog checks for console.log in modified JS/TS files.
func CheckConsoleLog(raw []byte) {
	if !gitutil.IsGitRepo() {
		hookio.Passthrough(raw)
		return
	}

	files := gitutil.ModifiedFiles([]string{`\.tsx?$`, `\.jsx?$`})

	// Filter out excluded patterns and non-existent files
	var filtered []string
	for _, f := range files {
		if !fileutil.Exists(f) {
			continue
		}
		excluded := false
		for _, pattern := range excludedPatterns {
			if pattern.MatchString(f) {
				excluded = true
				break
			}
		}
		if !excluded {
			filtered = append(filtered, f)
		}
	}

	hasConsole := false
	for _, file := range filtered {
		content, ok := fileutil.ReadFile(file)
		if ok && strings.Contains(content, "console.log") {
			hookio.Logf("[Hook] WARNING: console.log found in %s", file)
			hasConsole = true
		}
	}

	if hasConsole {
		hookio.Log("[Hook] Remove console.log statements before committing")
	}

	hookio.Passthrough(raw)
}

File: src/hooks/check_console_log.rs
//! check-console-log hook — scan modified JS/TS files for console.log.
//!
//! Ports internal/hooks/checkconsolelog.go.

use crate::{fileutil, gitutil, hookio};
use regex::Regex;
use std::path::Path;

pub fn check_console_log(raw: Vec<u8>) {
    if !gitutil::is_git_repo() {
        hookio::passthrough(&raw);
        return;
    }

    let files = gitutil::modified_files(&[r"\.tsx?$", r"\.jsx?$"]);

    let excluded = excluded_patterns();

    let filtered: Vec<String> = files
        .into_iter()
        .filter(|f| Path::new(f).exists() && !is_excluded(f, &excluded))
        .collect();

    let mut has_console = false;
    for file in &filtered {
        if let Some(content) = fileutil::read_file(Path::new(file)) {
            if content.contains("console.log") {
                hookio::log(&format!("[Hook] WARNING: console.log found in {file}"));
                has_console = true;
            }
        }
    }

    if has_console {
        hookio::log("[Hook] Remove console.log statements before committing");
    }

    hookio::passthrough(&raw);
}

fn excluded_patterns() -> Vec<Regex> {
    [
        r"\.test\.[jt]sx?$",
        r"\.spec\.[jt]sx?$",
        r"\.config\.[jt]s$",
        r"scripts/",
        r"__tests__/",
        r"__mocks__/",
    ]
    .iter()
    .map(|p| Regex::new(p).expect("valid static regex"))
    .collect()
}

pub fn is_excluded(path: &str, patterns: &[Regex]) -> bool {
    patterns.iter().any(|p| p.is_match(path))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn excludes_test_files() {
        let p = excluded_patterns();
        assert!(is_excluded("foo.test.ts", &p));
        assert!(is_excluded("foo.test.tsx", &p));
        assert!(is_excluded("foo.test.js", &p));
    }

    #[test]
    fn excludes_spec_and_config() {
        let p = excluded_patterns();
        assert!(is_excluded("foo.spec.ts", &p));
        assert!(is_excluded("vite.config.ts", &p));
        assert!(is_excluded("jest.config.js", &p));
    }

    #[test]
    fn excludes_tests_and_mocks_dirs() {
        let p = excluded_patterns();
        assert!(is_excluded("src/__tests__/x.ts", &p));
        assert!(is_excluded("src/__mocks__/x.ts", &p));
        assert!(is_excluded("scripts/run.ts", &p));
    }

    #[test]
    fn does_not_exclude_source_files() {
        let p = excluded_patterns();
        assert!(!is_excluded("src/app.ts", &p));
        assert!(!is_excluded("lib/helper.js", &p));
        assert!(!is_excluded("component.tsx", &p));
    }
}
